### PR TITLE
Run the multi-threaded executor at the end of each system task.

### PR DESCRIPTION
**Pull request by chescock**
# Objective

The multi-threaded executor currently runs in a dedicated task on a single thread.  When a system finishes running, it needs to notify that task and wait for the thread to be available and running before the executor can process the completion.  

See #8304

## Solution

Run the multi-threaded executor at the end of each system task.  This allows it to run immediately instead of needing to wait for the main thread to wake up.  Move the mutable executor state into a separate struct and wrap it in a mutex so it can be shared among the worker threads.  

While this should be faster in theory, I don't actually know how to measure the performance impact myself.

### CODE REVIEW

**Comment by NthTensor** [comment]
I will profile this later today. Are you sure the formatting of your SAFETY comments is correct?

**Comment by hymm** [comment]
ci failure on run-examples looks real.

**Comment by chescock** [comment]
@NthTensor 
> I will profile this later today. 

Thanks, but it's not actually working yet, so no need just yet.  The CI caught a race condition I had missed.  

> Are you sure the formatting of your SAFETY comments is correct?

Nope, not at all!  

... Yup, the ones in `should_run` are definitely wrong.  I put the new code in between the old SAFETY comment and the `unsafe` function call.  I was confused because I didn't see an `unsafe` keyword there, but I guess it was all in an `unsafe` function.  

I'll fix those.  Were those the ones you meant, or are there other issues, too?

**Comment by chescock** [comment]
Okay, I fixed the bug the CI caught, and added a unit test for it!  

("Dependencies / check-bans" seems to be failing for every PR and doesn't mention anything I changed, so I plan to ignore that failure.)

**Comment by hymm** [comment]
Seems promising. Ran the many foxes example with `RUST_LOG=bevy_ecs::schedule::schedule=info,bevy_app=info,bevy_render=info`. This gets rid of the most of the spans smaller than the schedule level.

Main schedule is saving 200us, extract is 70us and render is around 70us. Over a 8ms frame time.

main
image
extract
image
render
image
 
I'm going to run more of the examples add see how things look. I think running 3d_scene will be interesting to see how the overhead is reduced.

File: crates/bevy_tasks/src/single_threaded_task_pool.rs [hunk 1]
```diff
@@ -202,7 +202,7 @@ impl FakeTask {
 /// For more information, see [`TaskPool::scope`].
 #[derive(Debug)]
 pub struct Scope<'scope, 'env: 'scope, T> {
-    executor: &'env async_executor::LocalExecutor<'env>,
```
Comment (hymm) [inline crates/bevy_tasks/src/single_threaded_task_pool.rs:205]: why did you need to change these lifetimes?

Comment (chescock) [reply crates/bevy_tasks/src/single_threaded_task_pool.rs:205]: The futures passed to `scope.spawn` may run the executor and spawn more futures, so they need `&scope`.  That only lives for `'scope`, which is shorter than `'env`.  

The change to this file makes the API match the one in `task_pool.rs`.  

(The difference between the API in `task_pool` and `single_threaded_task_pool` led to some fun and confusing compiler errors.  I somehow had rust-analyzer using a configuration with `task_pool` but `cargo build` using a configuration with `single_threaded_task_pool`, so it worked in the IDE but failed on the command line.  And the lifetime errors it reports don't point at code in `single_threaded_task_pool`, so it took me a while to figure out that this file even existed.)

**Review by james7132 — COMMENTED** [review]
A more complete review is forthcoming, I'm particularly concerned about the soundness of the changes, so I may need some time to more thoroughly go through it.

With that said, I am seeing similar performance gains to what @hymm is seeing, so this is definitely looking pretty promising by itself.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 2]
```diff
@@ -483,12 +555,14 @@ impl MultiThreadedExecutor {
             }
 
             // Evaluate the system set's conditions.
+            // SAFETY: We have exclusive access to ExecutorState, so no other thread is touching these conditions
```
Comment (james7132) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:558]: This is not guaranteed given the function signature and the safety invariants, this likely needs to be propagated through the invariants of this function.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 3]
```diff
@@ -218,67 +241,38 @@ impl SystemExecutor for MultiThreadedExecutor {
...
-                #[cfg(feature = "trace")]
-                let mut executor = executor.instrument(executor_span);
-
-                // Immediately poll the task once to avoid the overhead of the executor
-                // and thread wake-up. Only spawn the task if the executor does not immediately
-                // terminate.
-                if block_on(poll_once(&mut executor)).is_none() {
-                    scope.spawn(executor);
-                }
+                let context = Context {
```
Comment (james7132) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:251]: Is it possible to merge this with SyncUnsafeSchedule? Both of which live for the duration of the execution.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 4]
```diff
@@ -522,132 +598,84 @@ impl MultiThreadedExecutor {
...
+        let world = unsafe { context.world_cell.world_mut() };
         // SAFETY: this system is not running, no other reference exists
-        let system = unsafe { &mut *systems[system_index].get() };
+        let system = unsafe { &mut *context.systems[system_index].get() };
+        // Move the full context object into the new future.
+        let context = *context;
+        #[cfg(feature = "trace")]
+        let system_span = self.system_task_metadata[system_index]
+            .system_task_span
+            .clone();
```
Comment (james7132) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:652]: Is this clone necessary?

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 5]
```diff
@@ -473,7 +545,7 @@ impl MultiThreadedExecutor {
         &mut self,
         system_index: usize,
         _system: &BoxedSystem,
-        conditions: &mut Conditions,
+        conditions: &Conditions,
```
Comment (james7132) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:548]: Is this 100% required to be read-only?

**Review by james7132 — COMMENTED** [review]
Generally looks good to me, though the potential for aliasing on `Conditions` is a bit concerning.

Ideally, we wouldn't yield back to the async executor or the OS, but we can leave that for another PR.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs (hunk 3 above)
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:251]: Yes, I think so.  I mainly left them separate to reduce the churn, and because I wasn't sure what to name things.  

A related issue is that `Context` gets copied into each task.  It holds a reference to the `scope`, so we have to construct the `Context` inside the closure passed to `scope_with_executor`, which means it doesn't live long enough to get passed to spawned tasks by reference.  

If we want to minimize the size of the tasks, then it would make sense to create one struct outside of the `scope_with_executor` call, so that we're only copying two pointers in each task.  

I'll try to come up with some better names and reorganize.  

(The other option is to get rid of the wrapping struct and pass `scope` as a separate reference, but then you need explicit lifetimes everywhere.)

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs (hunk 4 above)
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:652]: I think so, because that gets sent to a spawned task, and the borrow of `system_task_metadata` won't live long enough.  I didn't change that; the previous code did the `clone()` in the call to `task.instrument` and I just moved the call here.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs (hunk 5 above)
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:548]: No, probably not.  

The issue is that those conditions come from the schedule, which is only available during `run`, so they couldn't get put in the `ExecutorState` behind the mutex.  Since they were ultimately boxed systems, I figured I could treat them like the other systems.  But, yeah, this is hard to prove safe.  We need *something* shared because every task needs a copy to run the executor with, but it can be better than this.  

Hmm, I could create a `Mutex<&mut Conditions>` for the duration of `run` and pass a shared reference to that alongside the other mutex, then lock one after the other.  Would that be cleaner?  That would also address your concerns about the SAFETY comments.

**Comment by chescock** [comment]
I added two commits to try to address @james7132's review comments, then rebased the whole thing to resolve merge conflicts.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 6]
```diff
@@ -111,12 +128,20 @@ pub struct MultiThreadedExecutor {
...
     /// When set, stops the executor from running any more systems.
     stop_spawning: bool,
 }
 
+/// References to data required by the executor.
+/// This is copied to each system task so that can invoke the executor when they complete.
+// These all need to outlive 'scope in order to be sent to new tasks,
+// and keeping them all in a struct means we can use lifetime elision.
+#[derive(Copy, Clone)]
+pub struct Context<'scope, 'env: 'scope, 'sys> {
```
Comment (james7132) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:140]: Does this need to be pub?

```suggestion
struct Context<'scope, 'env: 'scope, 'sys> {
```

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 7]
```diff
@@ -208,69 +234,34 @@ impl SystemExecutor for MultiThreadedExecutor {
             .map(|e| e.0.clone());
         let thread_executor = thread_executor.as_deref();
 
-        let SyncUnsafeSchedule {
-            systems,
-            mut conditions,
-        } = SyncUnsafeSchedule::new(schedule);
+        let environment = &Environment::new(self, schedule, world);
```
Comment (james7132) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:237]:
```suggestion
        let environment = Environment::new(self, schedule, world);
```
Is this borrow needed?

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs (hunk 4 above)
Comment (james7132) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:650]: OK sounds good to me.

**Review by james7132 — COMMENTED** [review]
This likely requires another check to ensure the performance hasn't degraded from the extra mutex, and we can come back and replace it with the old SyncUnsafeCells if we can reasonably prove out the safety invariants, but this otherwise looks good to me.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs (hunk 7 above)
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:237]: `Context` needs a borrow because `Environment` owns the `Mutex<Conditions>`, and so that we only copy one pointer into each task instead of the whole environment.  And doing the borrow here avoids having to spell out `environment: &environment` when constructing the `Context`.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 8]
```diff
@@ -279,16 +270,63 @@ impl SystemExecutor for MultiThreadedExecutor {
...
+
+impl<'scope, 'env: 'scope, 'sys> Context<'scope, 'env, 'sys> {
+    fn system_completed(
+        &self,
+        system_index: usize,
+        res: Result<(), Box<dyn Any + Send>>,
+        system: &BoxedSystem,
+    ) {
+        // tell the executor that the system finished
+        self.environment
```
Comment (james7132) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:294]: This doesn't have to be done in this PR, but we have the SystemResult here, and we may be able to lock the executor, sending this through the completion queue when we can just pass it in via a function argument may be wasteful and may contribute to contention on the queue.

Perhaps we could only add the result to the queue if and only if we failed to get the lock on the executor state.

**Review by james7132 — APPROVED** [review]
Checked this against main, the new mutex does seem to eat into the improvements we saw earlier. Tested on many_foxes. Yellow is this PR, red is main.

main
image

render
image

There are still notable gains however. As @hymm noted, this most heavily impacts schedules with primarily small systems with little work to do.

First
image


PreUpdate
image

Last
image


What's interesting here is that there's almost 4x the number of calls into the multithreaded executor, but significantly improved the overall distribution of the time spent. On average cutting the time spent by 50%, and eliminating most of the long tail instances that take 100+us per run.
image

---

Overall, LGTM. We can try to remove the mutex and see if we can make any improvements 
in a follow-up PR, but I'm interested in trying to minimize yielding to the task executor.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs (hunk 8 above)
Comment (james7132) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:294]: Nevermind, just tried this myself, seems to deadlock.

**Comment by chescock** [comment]
Rebased to fix merge conflicts.

> the new mutex does seem to eat into the improvements we saw earlier.

Oh, wow, I wasn't expecting an uncontended mutex to actually cost anything!  One option to get back to one mutex with only safe code to would be to create a `Mutex<(&mut ExecutorState, Conditions<'_>)>`, but you'd have to pull the rest of `ExecutorState` into another struct to split the borrow.  (It doesn't sound like that's worth doing in this PR, though.)

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs (hunk 8 above)
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:294]: Yeah, there's a race condition if one thread fails to get the lock, then the other thread exits the executor, and then the first thread pushes to the queue.

**Comment by james7132** [comment]
> Oh, wow, I wasn't expecting an uncontended mutex to actually cost anything! 

Even if it's not under contention, there is still going to be a syscall to make sure that on the OS side there is no issue.

> It doesn't sound like that's worth doing in this PR, though.

Agreed. These are wins even with the impact from the extra mutex, and we can incrementally improve the results in a later PR.

File: crates/bevy_tasks/src/single_threaded_task_pool.rs (hunk 1 above)
Comment (hymm) [reply crates/bevy_tasks/src/single_threaded_task_pool.rs:205]: Might be worth pulling out of this pr then. This is a riskier change and if for some reason it gets reverted we should still keep the changes to this file.

**Review by hymm — COMMENTED** [review]
Nice job. The changes are pretty conservative since it's mostly just yeeting most of the logic behind a mutex. I took the opportunity to do a pass over the safety comments. Some of them weren't quite correct before.

One thing to note is that I think this will deadlock when there aren't any taskpool threads now, since we're not awaiting in the ticking code anymore. Not sure it's relevant, but should maybe be added to the migration guide.

I want to do a little more pref testing with varying thread counts before approving. Should get to that this weekend.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 9]
```diff
@@ -208,69 +234,34 @@ impl SystemExecutor for MultiThreadedExecutor {
...
+                context.tick_executor();
             },
         );
 
-        if self.apply_final_deferred {
+        // End the borrows of self and world in environment by copying out the reference to systems.
+        let systems = environment.systems;
+
+        let state = self.state.get_mut().unwrap();
+        if state.apply_final_deferred {
```
[the remark this answers is not in the document]
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:255]: Sure, I'll move that.  

(I'm surprised that's the only one; I wasn't looking at whether the fields were used mutably, just whether they were read outside the lock.  I figured there would be something else immutable in there, but even `num_systems` gets changed.)

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 10]
```diff
@@ -279,16 +270,63 @@ impl SystemExecutor for MultiThreadedExecutor {
...
+            .unwrap_or_else(|error| unreachable!("{}", error));
+        if let Err(payload) = res {
+            eprintln!("Encountered a panic in system `{}`!", &*system.name());
+            // set the payload to propagate the error
+            {
+                let mut panic_payload = self.environment.executor.panic_payload.lock().unwrap();
+                *panic_payload = Some(payload);
+            }
+        }
+        self.tick_executor();
```
[the remark this answers is not in the document]
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:310]: No, it spawns new tasks for each system and returns.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 11]
```diff
@@ -317,77 +364,114 @@ impl MultiThreadedExecutor {
...
-        systems: &'scope [SyncUnsafeCell<BoxedSystem>],
-        conditions: &mut Conditions,
-        world_cell: UnsafeWorldCell<'scope>,
-    ) {
+    unsafe fn spawn_system_tasks(&mut self, context: &Context<'_, '_, '_>) {
         if self.exclusive_running {
             return;
         }
 
+        let mut conditions = context
```
[the remark this answers is not in the document]
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:399]: That should be possible.  @james7132 said "We can try to remove the mutex and see if we can make any improvements
in a follow-up PR", so I'm not planning to change that in this PR.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 12]
```diff
@@ -317,77 +364,114 @@ impl MultiThreadedExecutor {
...
 
-            // SAFETY: `can_run` returned true, which means that:
-            // - It must have called `update_archetype_component_access` for each run condition.
-            // - There can be no systems running whose accesses would conflict with any conditions.
-            if unsafe { !self.should_run(system_index, system, conditions, world_cell) } {
-                self.skip_system_and_signal_dependents(system_index);
-                continue;
-            }
+            ready_systems.clear();
+            ready_systems.union_with(&self.ready_systems);
```
[the remark this answers is not in the document]
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:415]: It's still necessary because `ready_systems` gets modified inside the loop by `skip_system_and_signal_dependents()` if a system is skipped.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 13]
```diff
@@ -516,132 +600,84 @@ impl MultiThreadedExecutor {
...
-        let sender = self.sender.clone();
-        let panic_payload = self.panic_payload.clone();
+        let system = unsafe { &mut *context.environment.systems[system_index].get() };
+        // Move the full context object into the new future.
+        let context = *context;
+
+        let system_meta = &self.system_task_metadata[system_index];
+
+        #[cfg(feature = "trace")]
+        let system_span = system_meta.system_task_span.clone();
```
[the remark this answers is not in the document]
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:612]: Oh, I see, I made this span less useful by pulling the sending of the completion message out of it.  Sorry; I was trying to exclude `tick_executor()` and I had put those in the same function.  

The existing code was doing a non-blocking `try_send` on the channel, which calls `ConcurrentQueue::push` and then does some bookkeeping.  So the thread wouldn't have parked before and won't park now.  It looks like [there is a busy-wait in `push`](https://github.com/smol-rs/concurrent-queue/blob/93ee058b7f07b2346fdd55af6a033165239ef477/src/bounded.rs#L150), though, so that may be slow under contention.  

Do you want any changes here?  I could take the `tick_executor()` call out of `system_completed()` and then put `system_completed()` back in the tracing span.  Or I could remove the span.

[the remark this answers is not in the document]
Comment (hymm) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:612]: I think it's fine to leave as is for now. If we're not seeing much different here between the system run time we should probably remove it later.

**Comment by hymm** [comment]
Ran this 3d scene and saw a somewhat ambiguous result for Render schedule. The other schedules seemed faster, but rendering is often the bottleneck so want to be carefull here.

image

I think the regression is just due to there no longer being the weird fast path hump, so we're probably ok here.

tested with `RUST_LOG=bevy_ecs::schedule::schedule=info,bevy_app=info,bevy_render=info cargo run --profile stress-test --example 3d_scene -F trace_tracy` which removes the system level tracing spans.

Looks like I was wrong about it deadlocking if there wasn't any compute threads. I added
```rust
        .add_plugins(DefaultPlugins.set(TaskPoolPlugin {
            task_pool_options: TaskPoolOptions {
                min_total_threads: 0,
                max_total_threads: 0,
                compute: TaskPoolThreadAssignmentPolicy {
                    min_threads: 0,
                    max_threads: 0,
                    percent: 0.0,
                },
                ..default()
            }
        }))
```
and it ran ok. I confirmed with tracy that there weren't any compute threads.

**Review by hymm — APPROVED** [review]

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs [hunk 14]
```diff
@@ -317,77 +364,114 @@ impl MultiThreadedExecutor {
...
-            // SAFETY:
-            // - No other reference to this system exists.
-            // - `can_run` has been called, which calls `update_archetype_component_access` with this system.
-            // - `can_run` returned true, so no systems with conflicting world access are running.
-            unsafe {
-                self.spawn_system_task(scope, system_index, systems, world_cell);
+                self.ready_systems.set(system_index, false);
+
+                // SAFETY: `can_run` returned true, which means that:
+                // - It must have called `update_archetype_component_access` for each run condition.
```
[the remark this answers is not in the document]
Comment (JoJoJet) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:439]: `update_archetype_component` access is used to validate that the world matches, so it needs to be called before `run_unsafe` if you haven't already verified that the world is the same one used to initialize the system.
